notify/impl.go

```diff
@@ -585,17 +585,27 @@ func (n *Slack) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
 	if err != nil {
 		return true, redactURL(err)
 	}
-	resp.Body.Close()
+	defer resp.Body.Close()
 
-	return n.retry(resp.StatusCode)
+	return n.retry(resp)
+}
+
+func slackErr(statusCode int, body io.Reader) error {
+	if body != nil {
+		if bs, err := ioutil.ReadAll(body); err == nil {
+			return fmt.Errorf("unexpected status code %v: %s", statusCode, string(bs))
+		}
+	}
+	return fmt.Errorf("unexpected status code %v", statusCode)
 }
 
-func (n *Slack) retry(statusCode int) (bool, error) {
+func (n *Slack) retry(resp *http.Response) (bool, error) {
```

Comment:
I would change the signature of the method to `retry(statusCode int, body io.Reader)` so we don't need the extra `slackErr()` function.

Reply:
@simonpasquier if I do it, I'll make the code not so clean because in this method i do:
```go
    return statusCode/100 == 5, slackErr(statusCode, resp.Body)
```
removing the `slackErr` probably I'll do something like that:
```go
func (n *Slack) retry(statusCode int, body io.Reader) (bool, error) {
	// Only 5xx response codes are recoverable and 2xx codes are successful.
	// https://api.slack.com/incoming-webhooks#handling_errors
	// https://api.slack.com/changelog/2016-05-17-changes-to-errors-for-incoming-webhooks
	if statusCode/100 != 2 {
		retry := statusCode/100 == 5
		if body != nil {
			if bs, err := ioutil.ReadAll(body); err == nil {
				return retry, fmt.Errorf("unexpected status code %v: %q", statusCode, string(bs))
			}
		}
		return retry, fmt.Errorf("unexpected status code %v", statusCode)
	}

	return false, nil
}
```
if it's ok for you I can chance the code

Reply:
You can do this to minimize the number of nested levels:

```go
if statusCode/100 == 2 {
    return false, nil
}
err := fmt.Errorf("unexpected status code %v", statusCode)
if body != nil {
    if bs, errRead := ioutil.ReadAll(body); errRead == nil {
        err = fmt.Errorf("%s: %q", err, string(bs))
    }
}
return statusCode/100 == 5, err
```

Reply:
Done